crates/client/rpc/src/lib.rs

```diff
@@ -220,9 +226,101 @@ where
     }
 }
 
+#[async_trait]
+impl<A, B, BE, G, C, P, H> StarknetRpcApiCommonFuncs for Starknet<A, B, BE, G, C, P, H>
+where
+    A: ChainApi<Block = B> + 'static,
+    B: BlockT,
+    P: TransactionPool<Block = B> + 'static,
+    BE: Backend<B> + 'static,
+    C: HeaderBackend<B> + BlockBackend<B> + StorageProvider<B, BE> + 'static,
+    C: ProvideRuntimeApi<B>,
+    C::Api: StarknetRuntimeApi<B> + ConvertTransactionRuntimeApi<B>,
+    G: GenesisProvider + Send + Sync + 'static,
+    H: HasherT + Send + Sync + 'static,
+{
+    async fn declare_txn_common(
+        &self,
+        transaction_inputs: DeclareTransactionCommonInput,
+    ) -> Option<(TransactionHash, ClassHash)> {
+        let best_block_hash = self.get_best_block_hash();
+
+        match transaction_inputs {
+            DeclareTransactionCommonInput::V1(transaction) => {
+                let (class_hash, tx_hash) = (transaction.class_hash(), transaction.tx_hash());
+
+                let current_block_hash = self.get_best_block_hash();
+                let contract_class = self
+                    .overrides
+                    .for_block_hash(self.client.as_ref(), current_block_hash)
+                    .contract_class_by_class_hash(current_block_hash, class_hash);
+
+                if let Some(contract_class) = contract_class {
+                    error!("Contract class already exists: {:?}", contract_class);
+                    // return Err(StarknetRpcApiError::ClassAlreadyDeclared.into());
+                }
+
+                let extrinsic =
+                    self.convert_tx_to_extrinsic(best_block_hash, AccountTransaction::Declare(transaction)).unwrap();
+
+                submit_extrinsic(self.pool.clone(), best_block_hash, extrinsic).await.ok()?;
+
+                Some((tx_hash, class_hash))
+            }
+            DeclareTransactionCommonInput::V0(declare_txn, program_vec, entrypoints, abi_length) => {
```

Comment:
what if we made a custom `DeclareV0Transaction` type which has all these fields? i think it might be cleaner/maintainable, wdyt?